File: BackEnd/Application/UseCases/TaskToDoUseCase.cs
using Application.DTOs;
using Application.Repositories;
using Domain.Entities;
using System.Threading.Tasks;

namespace Application.UseCases;

public class TaskToDoUseCase
{
    private readonly ICrudDefault<TaskToDo, TaskToDoDto> _taskToDoAdapter;

    /// <summary>
    /// Initializes a new instance of the <see cref="TaskToDoUseCase"/> class.
    /// </summary>
    /// <param name="taskToDoAdapter">The adapter for performing CRUD operations on TaskToDo entities.</param>
    public TaskToDoUseCase(ICrudDefault<TaskToDo, TaskToDoDto> taskToDoAdapter)
    {
        _taskToDoAdapter = taskToDoAdapter;
    }

    /// <summary>
    /// Retrieves the TaskToDo item with the specified ID.
    /// </summary>
    /// <param name="id">The ID of the TaskToDo item to retrieve.</param>
    /// <returns>The TaskToDo item with the specified ID.</returns>
    public async Task<TaskToDo> Get(int id)
    {
        return await this._taskToDoAdapter.List(id);
    }

    /// <summary>
    /// Creates a new TaskToDo item with the specified details.
    /// </summary>
    /// <param name="taskToDoDto">A DTO containing the details for the new TaskToDo item.</param>
    /// <returns>The newly created TaskToDo item.</returns>
    public async Task<TaskToDo> Create(TaskToDoDto taskToDoDto)
    {
        var taskToDo = new TaskToDo
        {
            Title = taskToDoDto.Title,
            Description = taskToDoDto.Description,
            IsCompleted = taskToDoDto.IsCompleted,
            CategoryId = taskToDoDto.CategoryId,
            UserId = taskToDoDto.UserId
        };
        return await this._taskToDoAdapter.Create(taskToDo);
    }

    /// <summary>
    /// Updates an existing TaskToDo item with the specified details.
    /// </summary>
    /// <param name="taskToDoDto">A DTO containing the updated details for the TaskToDo item.</param>
    /// <returns>The updated TaskToDo item.</returns>
    public async Task<TaskToDo> Update(TaskToDoDto taskToDoDto)
    {
        var taskToDo = new TaskToDo
        {
            Id = taskToDoDto.Id,
            Title = taskToDoDto.Title,
            Description = taskToDoDto.Description,
            IsCompleted = taskToDoDto.IsCompleted,
            CategoryId = taskToDoDto.CategoryId,
            UserId = taskToDoDto.UserId
        };
        return await this._taskToDoAdapter.Update(taskToDo);
    }

    /// <summary>
    /// Deletes the TaskToDo item with the specified ID.
    /// </summary>
    /// <param name="id">The ID of the TaskToDo item to delete.</param>
    /// <returns>A <see cref="Task"/> representing the asynchronous operation.</returns>
    public async Task Delete(int id)
    {
        var taskToDo = await this._taskToDoAdapter.List(id);
        await this._taskToDoAdapter.Delete(taskToDo);
    }

    /// <summary>
    /// Retrieves all TaskToDo items for the specified user ID.
    /// </summary>
    /// <param name="id">The user ID to retrieve TaskToDo items for.</param>
    /// <param name="filter"></param>
    /// <returns>A list of TaskToDo items for the specified user.</returns>
    public async Task<List<TaskToDo>> GetAll(int id, int? filter = null)
    {
        return await this._taskToDoAdapter.GetAll(id, filter == null ? 0 : filter );
    }
}

File: frontend/src/Application/UseCases/TaskToDo.UseCase.ts
import ICrudDefault from "../../Domain/Repositories/ICrudDefault.Repository";

export default class TaskToDoUseCase {
    private taskToDoRepository: ICrudDefault;

    /// <summary>
    /// Initializes a new instance of the <see cref="TaskToDoUseCase"/> class.
    /// </summary>
    /// <param name="taskToDoRepository">The task to do repository.</param>
    constructor(taskToDoRepository: ICrudDefault) {
        this.taskToDoRepository = taskToDoRepository;
    }

    /**
     * Retrieves all tasks from the task repository, optionally filtered by the provided object.
     *
     * @param filter - An object containing the filter criteria to apply to the task retrieval.
     * @returns A Promise that resolves to an array of tasks matching the filter criteria.
     */
    async GetAll(filter: object): Promise<any> {
        return this.taskToDoRepository.GetAll(filter);
    }

    /**
     * Deletes a task from the task repository.
     *
     * @param id - The ID of the task to be deleted.
     * @returns A Promise that resolves when the task has been deleted.
     */
    async DeleteTask(id: number) {
        return this.taskToDoRepository.Delete(id);
    }

    /**
     * Adds a new task to the task repository.
     *
     * @param task - The task object to be added. All fields except "id" are required and cannot be null, undefined, or empty.
     * @returns A Promise that resolves to the newly created task.
     * @throws Error if any required field is missing or empty.
     */
    async AddTask(task: any) {
        for (const key in task) {
            if(key == "id") {
                continue;
            }
            
            if ([null, undefined, ''].includes(task[key])) {
                throw new Error('All fields are required');
            }
        }
        
        return this.taskToDoRepository.Create(task);
    }

    /**
     * Updates a task in the task repository.
     *
     * @param task - The task object to be updated. All fields except "user" are required and cannot be null, undefined, or empty.
     * @returns A Promise that resolves to the updated task.
     * @throws Error if any required field is missing or empty.
     */
    async UpdateTask(task: any) {
        console.log(task)
        for (const key in task) {
            if(["user", "createdAt", "updateddAt", "category"].includes(key)) {
                continue;
            }
            
            if ([null, undefined, ''].includes(task[key])) {
                throw new Error('All fields are required');
            }
        }
        
        return this.taskToDoRepository.Update(task);
    }
}
